GCC 12.0 false positive `-Wreturn-type` warning on some uses of `ERR_FAIL_V_MSG`

### Godot version

3.x (eb3f0f50b9), master (3078b92d)

### System information

Linux, Mageia 9 x86_64

### Issue description

My distro (dev version) just updated to a pre-release snapshot of GCC 12 (`gcc (Mageia 12.0.1-0.20220227.2.mga9) 12.0.1 20220227 (experimental)`), and I'm seeing a regression with the `-Wreturn-type` warning.

It's not a Godot bug, but triggered by our code, so I'm document it here for now until I / others can manage to isolate it in a minimal reproduction project to report the bug to GCC upstream.

Compiling the current `master` branch (same in `3.x`) with the following command:
```
scons -j7 p=linuxbsd verbose=yes warnings=extra werror=no tests=yes
```
I get the following warnings:
```
platform/linuxbsd/os_linuxbsd.cpp: In member function 'virtual String OS_LinuxBSD::get_processor_name() const':
platform/linuxbsd/os_linuxbsd.cpp:156:1: warning: control reaches end of non-void function [-Wreturn-type]
  156 | }
      | ^
```
```
drivers/unix/file_access_unix.cpp: In member function 'virtual uint32_t FileAccessUnix::_get_unix_permissions(const String&)':
drivers/unix/file_access_unix.cpp:326:1: warning: control reaches end of non-void function [-Wreturn-type]
  326 | }
      | ^
```
```
platform/osx/export/codesign.cpp: In static member function 'static Error CodeSign::codesign(bool, bool, const String&, const String&, String&)':
platform/osx/export/codesign.cpp:1567:1: warning: control reaches end of non-void function [-Wreturn-type]
 1567 | }
      | ^
```
```
platform/osx/export/plist.cpp: In member function 'bool PList::load_file(const String&)':
platform/osx/export/plist.cpp:366:1: warning: control reaches end of non-void function [-Wreturn-type]
  366 | }
      | ^
```
```
editor/editor_resource_preview.cpp: In member function 'virtual Ref<Texture2D> EditorResourcePreviewGenerator::generate(const RES&, const Size2&) const':
editor/editor_resource_preview.cpp:57:1: warning: control reaches end of non-void function [-Wreturn-type]
   57 | }
      | ^
```

And some more like these (I found 21 warnings total in `master`). They're all cases where we return with `ERR_FAIL_V_MSG`, e.g.:

```cpp
Ref<Texture2D> EditorResourcePreviewGenerator::generate(const RES &p_from, const Size2 &p_size) const {
	Ref<Texture2D> preview;
	if (GDVIRTUAL_CALL(_generate, p_from, p_size, preview)) {
		return preview;
	}
	ERR_FAIL_V_MSG(Ref<Texture2D>(), "EditorResourcePreviewGenerator::_generate needs to be overridden.");
}
```

But `ERR_FAIL_V_MSG` should always return, so GCC is wrong:
```cpp
/**
 * Try using `ERR_FAIL_COND_V_MSG`.
 * Only use this macro if more complex error detection or recovery is required.
 *
 * Prints `m_msg`, and the current function returns `m_retval`.
 */
#define ERR_FAIL_V_MSG(m_retval, m_msg)                                                                                  \
	if (true) {                                                                                                          \
		_err_print_error(FUNCTION_STR, __FILE__, __LINE__, "Method/function failed. Returning: " _STR(m_retval), m_msg); \
		return m_retval;                                                                                                 \
	} else                                                                                                               \
		((void)0)
```
(The `else ((void)0)` is a trick to force compilers to error when the macro is used without a semicolon - just for code consistency.)

It also doesn't raise this warning in all cases, for example that same class has this very similar method which doesn't trigger the warning:
```cpp
bool EditorResourcePreviewGenerator::handles(const String &p_type) const {
	bool success;
	if (GDVIRTUAL_CALL(_handles, p_type, success)) {
		return success;
	}
	ERR_FAIL_V_MSG(false, "EditorResourcePreviewGenerator::_handles needs to be overridden.");
}
```

I tried to reproduce the issue in a minimal project, but that didn't work. Here's my attempt:
```cpp
#include <cstdio>

#define ERR_FAIL_V_MSG(m_retval, m_msg) \
	if (true) { \
		printf("%s\n", m_msg); \
		return m_retval; \
	} else \
		((void)0)

int func() {
	ERR_FAIL_V_MSG(42, "Got an error!");
}

int main() {
	int ret = func();
	printf("%d\n", ret);
	return 0;
}
```

CC @marxin if you can help reduce this to something useful for upstream.

### Steps to reproduce

- Compile Godot with a recent snapshot of GCC 12 (note that GCC 12 is not released yet, it's currently in feature freeze and regression fixing time) with `warnings=extra werror=yes` to abort on first error.

### Workaround

Compile with `CCFLAGS="-Wno-return-type"` added to the `scons` command.

Here's one corresponding to:
```
platform/linuxbsd/os_linuxbsd.cpp: In member function 'virtual String OS_LinuxBSD::get_processor_name() const':
platform/linuxbsd/os_linuxbsd.cpp:156:1: warning: control reaches end of non-void function [-Wreturn-type]
  156 | }
      | ^
```

Build command:
```
g++ -E -o platform/linuxbsd/os_linuxbsd.linuxbsd.tools.64.o -c -std=gnu++17 -Wctor-dtor-privacy -Wnon-virtual-dtor -Wplacement-new=1 -Wno-error=cpp -g3 -pipe -Wall -Wextra -Wwrite-strings -Wno-unused-parameter -Wshadow-local -Wno-misleading-indentation -Walloc-zero -Wduplicated-branches -Wduplicated-cond -Wstringop-overflow=4 -Wlogical-op -Wattribute-alias=2 -Werror -DDEBUG_ENABLED -DDEV_ENABLED -DNO_EDITOR_SPLASH -DTOUCH_ENABLED -DALSA_ENABLED -DALSAMIDI_ENABLED -DPULSEAUDIO_ENABLED -D_REENTRANT -DDBUS_ENABLED -DJOYDEV_ENABLED -DUDEV_ENABLED -DX11_ENABLED -DUNIX_ENABLED -D_FILE_OFFSET_BITS=64 -DVULKAN_ENABLED -DGLES3_ENABLED -DTOOLS_ENABLED -DMINIZIP_ENABLED -DZSTD_STATIC_LINKING_ONLY -DUSE_VOLK -DVK_USE_PLATFORM_XLIB_KHR -DGLAD_ENABLED -DGLES_OVER_GL -DHAVE_MNTENT -Ithirdparty/freetype/include -Ithirdparty/libpng -Ithirdparty/glad -Ithirdparty/volk -Ithirdparty/vulkan -Ithirdparty/vulkan/include -Ithirdparty/zstd -Ithirdparty/zlib -Iplatform/linuxbsd -I. -I/usr/include/dbus-1.0 -I/usr/lib64/dbus-1.0/include platform/linuxbsd/os_linuxbsd.cpp
```

(XZ compressed, and zipped so that GitHub would let me upload it.)

[os_linuxbsd.preprocessed.log.xz.zip](https://github.com/godotengine/godot/files/8186289/os_linuxbsd.preprocessed.log.xz.zip)

--- a/SConstruct
+++ b/SConstruct
@@ -618,10 +618,14 @@ if selected_platform in platform_list:
                 env.Append(CXXFLAGS=["-Wno-error=cpp"])
                 if cc_version_major == 7:  # Bogus warning fixed in 8+.
                     env.Append(CCFLAGS=["-Wno-error=strict-overflow"])
+                if cc_version_major >= 12:  # False positives in our error macros, see GH-58747.
+                    env.Append(CCFLAGS=["-Wno-error=return-type"])
             elif methods.using_clang(env) or methods.using_emcc(env):
                 env.Append(CXXFLAGS=["-Wno-error=#warnings"])
-        else:  # always enable those errors
-            env.Append(CCFLAGS=["-Werror=return-type"])
+        else:  # Always enable those errors.
+            # False positives in our error macros, see GH-58747.
+            if not (methods.using_gcc(env) and cc_version_major >= 12):
+                env.Append(CCFLAGS=["-Werror=return-type"])
 
     if hasattr(detect, "get_program_suffix"):
         suffix = "." + detect.get_program_suffix()

